File: src/mem/ruby/structures/Scratchpad.py
from m5.params import *
from m5.SimObject import SimObject


class Scratchpad(SimObject):
    type = "Scratchpad"
    cxx_header = "mem/ruby/structures/scratchpad.hh"
    cxx_class = "gem5::ruby::Scratchpad"

    time_to_wait = Param.Latency("Time before firing the event")
    size = Param.MemorySize("64MiB", "Size of memory inside of scratchpad")
    ruby_system = Param.RubySystem(Parent.any, "the parent ruby system")

File: src/mem/ruby/structures/Scratchpad.cc
#include "learning_gem5/part2/scratchpad.hh"

#include "base/trace.hh"
#include "debug/HelloExample.hh"

namespace gem5
{

Scratchpad::Scratchpad(ScratchpadParams *params) :
    SimObject(params),
	event([this]{processEvent();}, name()),
	myName(params.name),
	latency(params.time_to_wait),
	size(params.size)
{
    DPRINTF(Scratchpad, "Created the scratchpad\n");
}

void
Scratchpad::startup()
{
	schedule(event, latency)
}

void
Scratchpad::processEvent()
{
	timesLeft--;
	DPRINTF(Scratchpad, "Processing Scratch event :)  %d left\n", timesLeft);

	if (timesLeft <= 0) {
		DPRINTF(Scratchpad, "Done with scratch \n");
	} else {
		schedule(event, curTick() + latency);
	}
}


} // namespace gem5
